src/server.c

```diff
@@ -163,10 +164,17 @@ server_daemonize(const char *pidfile) {
 
 	/* write pidfile */
 	if (pidfile) {
-		FILE *f = fopen(pidfile, "w");
-		if (f) {
-			fprintf(f, "%d\n", (int)getpid());
-			fclose(f);
+		int pid_fd = open(pidfile, O_WRONLY | O_CREAT, 0600);
+		if (pid_fd > 0) {
+			char pid_buffer[(CHAR_BIT * sizeof(int) / 3) + 3]; /* max length for int */
+			int pid_sz = snprintf(pid_buffer, sizeof(pid_buffer), "%d\n", (int)getpid());
+			size_t written;
+			int written_total = 0;
+			while ((written = write(pid_fd, pid_buffer + written_total, pid_sz - written_total)) > 0
+				&& written_total < pid_sz) {
+				written_total += written;
+			}
+			close(pid_fd);
 		}
```

Comment:
we should log if we can't write the pid. Also, `write(2)` returns `ssize_t` (not `size_t`) and I'd suggest adding `O_TRUNC` and `O_NOFOLLOW`.